### DISCUSSION TITLE
tutorials/lift-off-part4/08-the-usemutation-hook

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#52] [course feedback]
# tutorials/lift-off-part4/08-the-usemutation-hook

https://www.apollographql.com/tutorials/lift-off-part4/08-the-usemutation-hook

**Comment by OliverFarren** [comment]
No mention of the 'options' object before the question on it?

**Reply by bljdavidson** [reply]
It does. About a 3rd of the way down.

>The first element is the mutate function we'll use to actually run the mutation later on. We'll call it incrementTrackViews. The second element is an object with information about the mutation: loading, error and data. This component doesn't need it, so we don't have to extract it.

Granted, it doesn't say this is the `options` object, but it should be fairly easy to figure it out from the question :)

**Reply by lizhennessy** [reply]
Thanks for calling this out, @OliverFarren ! We've made an update to the course to help clarify this. The `options` object is the second parameter that is passed into the `useMutation` function. The first is the name of the mutation operation itself, while the second is the `options` object that holds any variables the mutation might need! Hopefully this clarification helps to resolve any confusion. Please let us know if you have any other questions!

**Reply by neogeek13** [reply]
Yeah, without the example of where exactly that goes I wasn't following. I'm not the ideal student here as TS is not something I've used often (primarily mobile) but was getting lost in all the words where one line of example code (even a "we're not actually using this but if we did" like you did in I think it was the 3rd course) would have clarified what I needed to know.

**Comment by jraldrin** [comment]
I worked thru the typescript tutorial and I'm confused about the functions in the resolver.ts file not having arg and return types.  Seems like an important file and not having that info (arg and return types) makes the file hard to work with.  I looked at the generated types (types.ts) and I kind of get a feel for how things are put together but it pretty complex.

**Reply by lizhennessy** [reply]
Hi there @jraldrin ! I totally agree; there's definitely a lot of complexity in the `types.ts` file. The `Resolvers` type that we import into our `resolvers.ts` file, for example, has all of the information about _everything_ that our `resolvers` object is responsible for. If you hover over one of the destructured `dataSources` arguments, you'll see that the object-level type annotation make it possible to see that `dataSources` is actually an object with a `trackAPI` property, which is a `TrackAPI` instance. So even though all of our argument types - and the function return types - are taken care of by this one broad `Resolvers` type that we import, it definitely leads to some of those details being hidden in the process. 

We could totally add those annotations inline within the `resolvers.ts` file instead - a helpful experiment is to modify a resolver, like `Query.tracksForHome`, and annotate it with a (incorrect) return type. We get a helpful error from the `codegen` package telling us that there's a mismatch between what the resolver function claims to return, and what the method it calls _actually_ returns. So this can help us peek under the hood to see where those types are coming from, and how they've been compiled based on the info we've fed into the `codegen.ts` file. 

<img width="942" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/13683d4b-023d-40e5-ba81-8573f1d71b97">

So that allows us to dig a bit deeper into all of the different types in `types.ts` and add the annotation ourselves. So for instance, we know that the `tracksForHome` resolver calls the `getTracksForHome()` method on our data source. If we hover over that method in `track-api.ts`, we'll see that it resolves to `Promise<TrackModel[]>`. So we can totally add that to the function in `resolvers`. (This one's just a bonus type annotation though!)

<img width="851" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/9f15ec48-6bb8-4a80-80a2-49774d16d2bf">

**Comment by GeorgeOrzanescu** [comment]
"The first element is the mutate function we'll use to actually run the mutation later on. We'll call it incrementTrackViews. "

In this quote in sound that i can choose the naming. But i see that is identical to the resolver function handling this. Is the same naming required ?

**Reply by mabuyo** [reply]
Great question @GeorgeOrzanescu! You can definitely name the function whatever you'd like, it does _not_ need to match what's defined in your schema or resolvers. However, I do think it makes it easier to search for if you choose the same name :)

**Comment by benifbond** [comment]
Why do I still having this error in the mutation operation ?

{
  "data": {
    "incrementTrackViews": {
      "code": 404,
      "success": false,
      "message": "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Cannot PATCH /track/c_1/views</pre>\n</body>\n</html>\n",
      "track": null
    }
  }
}

**Reply by mabuyo** [reply]
Hey there @benifbond - your endpoint looks incorrect, can you double check the TrackAPI data source in `server/src/datasources/track-api.js`?

```
   return this.patch(`track/${trackId}/numberOfViews`);
```

It should be `numberOfViews`, not `views`.

**Comment by SimDav888** [comment]
Excellent tutorial again. Thank a lot for your hardwork!

**Comment by Souravsinha96** [comment]
in video we can see that we import gql from @apollo/client
but in theory it is imported from import { gql } from "../__generated__";
which one is correct??

**Reply by lizhennessy** [reply]
Hi there @Souravsinha96 ! Right now the Lift-off course videos are unfortunately only available following the _JavaScript_ version of the code. If you're on the TypeScript version, please disregard the JS import and use the following instead!

```import { gql } from "../generated"```

The course content for the TypeScript version will be the most accurate for you to reference. :)

**Reply by Souravsinha96** [reply]
oh!, My Bad 😅

**Comment by jstmpelowycz** [comment]
Hi, guys! What an excellent and easy-to-grasp tutorial!

Just a thing to point out — I believe, the views incrementation function should've been placed to `track.tsx`, sending it on initial page mounting (i.e. in `useEffect`).

However, you might've make it your own way to keep things simple, not making some people scared of the components effects.

Sorry for being picky.

Anyway, what you're doing is great, many thanks to the team!

**Reply by mabuyo** [reply]
Thanks so much for the feedback, appreciate it! And yes, that totally makes sense to increment views when the track page loads up the first time. The onClick handler was just easier to showcase without delving into more React-specific code!

**Comment by ghost** [comment]
Hello,
I am getting this error when I run "npm run generate" . It gives me red line below INCREMENT_TRACK_VIEWS saying the type is unknown.

const [ incrementTrackViews 1 = useMutation INCREMENT_TRACK_VIEWS, {
variables: { incrementTrackViewsId: id}

const INCREMENT_TRACK_VIEWS: unknown

ERROR:
" Argument of type 'unknown' is not assignable to parameter of type 'DocumentNode | TypedDocumentNode<any, OperationVariables>'.ts(2345) "

Please help.

**Reply by ghost** [reply]
I'm using Typescript version, importing gql from "../__generated__"

**Reply by ghost** [reply]
EDIT: I got the problem solved. It was an error in my side.

**Reply by mabuyo** [reply]
Glad to hear you were able to troubleshoot and fix it @Bhavshu -- let us know if you run into any other issues!

**Reply by alx168** [reply]
Hi there, I'm having this same issue. @Bhavshu or @mabuyo, could you please explain why this might be happening and the steps taken to fix it?

**Reply by alx168** [reply]
whoops, actually it was because I was running the 'npm run generate' command in my server folder. If you're running into this issue, switch to the client folder then run the command.

**Reply by mabuyo** [reply]
Glad to hear you were able to fix @alx168 !

**Comment by stephan-buckmaster** [comment]
An alternative design would be to trigger the increment-track-views from the server as opposed to the client.
However, I get the impression, while it would be possible it would not be standard to add this to the resolver code.
Is there another way? Or would that be proper?

**Reply by lizhennessy** [reply]
Hi there @stephan-buckmaster ! That's an interesting concept: would you consider the page load, or request from the server, to be the triggering event? I suppose that could work if each page were served up separately, rather than as a single-page app as we've built here. I would say that for the most part our resolver functions should be really focused without side effects. The nice thing about triggering the mutation from the frontend is that our mutating operation has a single direct effect on the underlying datasource - incrementing the track views. If we were to reserve the responsibility of incrementing track views somewhere else on our server, then I wouldn't include that logic in the resolver function, as we would no longer be calling that mutation operation from the frontend of our application. It's good to think about the GraphQL schema as the contract between the frontend and the backend; where the frontend client is responsible for calling the queries and mutations!
